hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitionsAsync.java

```diff
@@ -0,0 +1,208 @@
+                                setPartitionResult(partitionId, t);
+                                decrementLatchAndHandle(1);
+                            }
+                        });
+    }
+
+    private void decrementLatchAndHandle(int count) {
+        if (latch.addAndGet(-count) > 0) {
+            // we're not done yet
+            return;
+        }
+
+        Map<Integer, Object> result = createHashMap(partitionResults.length());
+        for (int partitionId = 0; partitionId < partitionResults.length(); partitionId++) {
+            Object partitionResult = partitionResults.get(partitionId);
+            if (partitionResult instanceof Throwable) {
+                future.setResult(partitionResult);
+                if (callback != null) {
+                    callback.onFailure((Throwable) partitionResult);
+                }
+                return;
+            }
+            // partitionResult is null for partitions which had no keys
+            if (partitionResult != null) {
+                result.put(partitionId, partitionResult);
+            }
+        }
+        future.setResult(result);
+        if (callback != null) {
+            callback.onResponse(result);
+        }
+    }
+
+    private class FirstAttemptExecutionCallback implements ExecutionCallback<Object> {
+        private final List<Integer> allPartitions;
+
+        FirstAttemptExecutionCallback(List<Integer> partitions) {
+            this.allPartitions = partitions;
+        }
+
+        @Override
+        public void onResponse(Object response) {
+            PartitionResponse result = operationService.nodeEngine.toObject(response);
+            Object[] results = result.getResults();
+            int[] partitions = result.getPartitions();
+            assert results.length <= allPartitions.size() : "results.length=" + results.length
+                    + ", but was sent to just " + allPartitions.size() + " partitions";
+            int failedPartitionsCnt = 0;
+            for (int i = 0; i < partitions.length; i++) {
+                if (results[i] instanceof Throwable) {
+                    retryPartition(partitions[i]);
+                    failedPartitionsCnt++;
+                } else {
+                    setPartitionResult(partitions[i], results[i]);
+                }
+            }
+            decrementLatchAndHandle(allPartitions.size() - failedPartitionsCnt);
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+            if (operationService.logger.isFinestEnabled()) {
+                operationService.logger.finest(t);
+            } else {
+                operationService.logger.warning(t.getMessage());
+            }
+            for (Integer partition : allPartitions) {
+                retryPartition(partition);
+            }
+        }
+    }
+
+    private void setPartitionResult(int partition, Object result) {
+        assert result != null : "null result";
```

Comment:
Is `null` a legitimate result when using the old `InvokeOnPartitions`?

Reply:
In general I think it is. In the specific case of `MultipleEntryOperation` it's not. I changed it to allow null.